arwen/contexts/output.go

```diff
@@ -237,7 +308,7 @@ func mergeOutputAccounts(
 		leftAccount.BalanceDelta = big.NewInt(0)
 	}
 	if rightAccount.BalanceDelta != nil {
-		leftAccount.BalanceDelta = big.NewInt(0).Add(leftAccount.BalanceDelta, rightAccount.BalanceDelta)
+		leftAccount.BalanceDelta = rightAccount.BalanceDelta
```

Comment:
Previously, `Add` was required indeed. This PR changes the meaning of the state stacks, from "partial states that can be merged" to actual "full checkpoints", with limited branching and revert (with a minor exception - storage updates, which still support merging). 

More info on how state stacking has changed: a good example is `executeOnSameContext()`, which must either finish successfuly, or all its changes reverted in case of failure, including any value transfers. This was solved by changing the contexts to hold an "active state" and a "state stack". Pushing on the state stack means copying the active state to the stack, while continuing to work with the active state. 

When `executeOnSameContext()` runs, it will modify the "active state" (with storage updates, `BalanceDelta` etc.). If it returns with success, the active state is considered "good", and the previous copy (at the top of the stack) is deleted with `PopDiscard()`. If it returns with failure, the active state is considered "wrong" and it is replaced with the copy from the top of the stack, with `PopSetActiveState()`.

The source line highlighted here in `output.go` is only reached by `executeOnDestContext()` on a successful return, when it must return to the previous SC on the stack. This requires a merge between the last known state of the previous SC (from the stack) with the _new_ active state produced by `executeOnDestContext()`. A simple replacement with `PopSetActiveState()` would not have been sufficient, like for `executeOnSameContext()`, because `executeOnDestContext()` copies it for the new SC (so that balances and nonces are visible to the new SC), but then censors it, so that the new SC appears to be running in isolation (almost everything is erased, except storage updates, which aren't accessible anyway). 

In conclusion, assignment is required here now, because `leftAccount.BalanceDelta` and `rightAccount.BalanceDelta` were initially identical, but `rightAccount.BalanceDelta` is the new value after `executeOnDestContext()`. 

This is tested by `TestExecution_ExecuteOnDestContext_Recursive_Mutual_SCs()`, with two SCs calling each other recursively using `executeOnDestContext()`.

While the current stacking concept makes sense and appears to work, it is not easy enough to understand from the code directly, and could be improved. I'm open to any idea.